frontend/src/Teacher/StudentTransfer.vue

```diff
@@ -0,0 +1,258 @@
+}
+
+interface Student {
+  username: string;
+  first_name: string;
+  last_name: string;
+}
+
+interface Class {
+  id: number;
+  teacher_username: string;
+  timeslot: string;
+  code: string;
+  subject_abbr: string;
+  csv_link: string;
+  assignments: Assignment[];
+  summary: string;
+  students: Student[];
+}
+
+interface Transfer {
+  submit_id: number;
+  before: string;
+  after: string;
+}
+
+interface AssignmentRequest {
+  name: string;
+  assigned_points: number | string;
+  src_assignment_id: number;
+  dst_assignment_id: number;
+}
+
+interface TransferRequest {
+  src_class: number;
+  dst_class: number;
+  student: string;
+  assignments: AssignmentRequest[];
+}
+
+const semester = ref(null);
+const subject = ref(null);
+const teacher = ref(null);
+const clazz = ref(null);
+const student = ref(null);
+const target_class = ref(null);
+const assignments = ref<AssignmentRequest[]>([]);
+const busy = ref(false);
+const data = await loadClasses();
+const classes = ref<Class[] | null>(null);
+const result = ref<Transfer[] | null>(null);
+
+const notAssignedCount = computed(
+  () => assignments.value.filter((t) => !t.dst_assignment_id).length
+);
+const uniqueAssignments = computed(() => true);
+const canTransfer = computed(
+  () =>
+    assignments.value.length &&
+    notAssignedCount.value === 0 &&
+    uniqueAssignments.value &&
+    !busy.value
+);
+
+async function onSubjectChanged() {
+  teacher.value = null;
+  clazz.value = null;
+  student.value = null;
+  target_class.value = null;
+  assignments.value = [];
+
+  const res = await getFromAPI<{ classes: Class[] }>(
+    `/api/classes?semester=${semester.value}&subject=${subject.value}`
+  );
+  if (res) {
+    classes.value = res.classes;
+  }
+}
+
+async function loadClasses(): Promise<Classes> {
+  return await getFromAPI<Classes>('/api/classes/all');
+}
+
+function target_changed() {
+  const target = classes.value.find((c) => c.code === target_class.value).assignments;
+  let index = 0;
+  assignments.value = classes.value
+    .find((c) => c.code === clazz.value)
+    .assignments.map((t) => ({
+      name: t.name,
+      assigned_points: t.students[student.value]?.assigned_points || '',
+      src_assignment_id: t.assignment_id,
+      dst_assignment_id: index < target.length ? target[index++].assignment_id : null
+    }));
+}
+
+async function transfer() {
+  busy.value = true;
+  const req: TransferRequest = {
+    student: student.value,
+    src_class: classes.value.find((c) => c.code === clazz.value).id,
+    dst_class: classes.value.find((c) => c.code === target_class.value).id,
+    assignments: assignments.value
+  };
+
+  const res = await getDataWithCSRF<{
+    transfers: Transfer[];
+  }>('/api/transfer_students', 'POST', req);
+
+  result.value = res.transfers;
+  busy.value = false;
+}
+</script>
+
+<template>
+  <div v-if="data">
+    <select v-model="semester">
+      <option v-for="item in Object.keys(data.semesters)" :key="item" :value="item">
```

Comment:
![image](https://github.com/user-attachments/assets/f929d4ba-5329-42f4-be78-561d967b4dd7)

The sorting isn't great here :) The keys should be sorted everywhere, they might be in arbitrary order.

Reply:
Will look into that.